### PR TITLE
feat: addition of codeflare stack to monitoring

### LINKED ISSUE #373: Add CodeFlare components as scrape targets and alerting rules
We require Prometheus to actively look for the pods and fire alerts if the conditions are met. CodeFlare and MCAD components need to expose metrics endpoint before alerts are passing. 

Alerts can be added to test if Prometheus is successfully scraping the endpoints and to meet with SLIs/SLOs

### PR DESCRIPTION
**Pull request by dimakis**
Closes #373 
## Description

For reference if needed: https://github.com/red-hat-data-services/odh-deployer/pull/365

The respective SOPs are on these merge requests:
https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/merge_requests/95
https://gitlab.cee.redhat.com/dsaridak/managed-tenants-sops/-/merge_requests/1 

We want the alerts' runbook/triage to point to their SOPs

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] For commits that came from upstream, `[UPSTREAM]` has been prepended to the commit message.
- [ ] JIRA link(s):
- [ ] The Jira story is acked.
- [ ] Live build image: 
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work.
- [ ] QE contact acknowledges that this has been tested and is approved for merge.

### CODE REVIEW

**Comment by ChristianZaccaria** [comment]
For reference if needed: https://github.com/red-hat-data-services/odh-deployer/pull/365

**Comment by dimakis** [comment]
@ChristianZaccaria can you verify this and stick up a couple of screen shots please?

**Comment by ChristianZaccaria** [comment]
> @ChristianZaccaria can you verify this and stick up a couple of screen shots please?


MCAD and KubeRay are not getting scraped at the moment. 

For MCAD, this has to do with [this PR for MCAD](https://github.com/project-codeflare/multi-cluster-app-dispatcher/pull/486). There are two things to this:

- It is exposing the endpoint on port `9090` at the moment, where I think you originally wanted to use `8083`.
- Something in that PR is causing MCAD to fail the deployment.

A similar issue could be for KubeRay.

image

**Comment by zdtsw** [comment]
unless i missed something, should not this be added into https://github.com/red-hat-data-services/odh-manifests/blob/master/monitoring/prometheus/prometheus-configs.yaml as well?
it wont be in use for v2 for now (since we have not done cloud-service part) 
we are not keep syncing changes from odh-deployer but mostly from odh-manifests, it might get lost somewhere in the near future when the common monitoring solution is moving into operator entierly

File: monitoring/prometheus/prometheus-configs.yaml [hunk 1]
```diff
@@ -745,6 +895,59 @@ data:
...
+      - name: Distributed Workloads CodeFlare
+        interval: 1m
+        rules:
+          - alert: CodeFlare Operator is not running
+            expr: absent(up{job=~'CodeFlare Operator'}) or up{job=~'CodeFlare Operator'} != 1
+            labels:
+              severity: warning
+            annotations:
+              description: This alert fires when the CodeFlare Operator is not running.
+              triage: 'https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/blob/main/RHODS/Distributed-Workloads/codeflare-operator-availability.md'
```
Comment (zdtsw) [inline monitoring/prometheus/prometheus-configs.yaml:845]: need an update on this

File: monitoring/prometheus/prometheus-configs.yaml [hunk 2]
```diff
@@ -927,6 +1136,63 @@ data:
...
+
+    - job_name: 'CodeFlare Operator'
+      honor_labels: true
+      metrics_path: /metrics
+      scheme: http
+      kubernetes_sd_configs:
+        - role: endpoints
+          namespaces:
+            names:
+              - openshift-operators
```
Comment (zdtsw) [inline monitoring/prometheus/prometheus-configs.yaml]: need double check this namespace.
prometheus is only applying to managed  service not on ODH

File: monitoring/prometheus/prometheus-configs.yaml (hunk 1 above)
Comment (ChristianZaccaria) [reply monitoring/prometheus/prometheus-configs.yaml:845]: Hi Wen, I made a MR for the outstanding triages: https://gitlab.cee.redhat.com/service/managed-tenants-sops/-/merge_requests/96 

With the CodeFlare Operator Redesign, where MCAD and InstaScale is embedded. The MCAD alert rules won't be necessary in this PR, they will need to be removed. For this same reason, I removed the MCAD triages in the mentioned MR.

Comment (zdtsw) [reply monitoring/prometheus/prometheus-configs.yaml:845]: Thanks @ChristianZaccaria 
Could you update this PR with the latest change (if any rules job etc are not valid)?
I will need the final version of what is correct for DW to work.

**Comment by ChristianZaccaria** [comment]
> unless i missed something, should not this be added into https://github.com/red-hat-data-services/odh-manifests/blob/master/monitoring/prometheus/prometheus-configs.yaml as well? it wont be in use for v2 for now (since we have not done cloud-service part) we are not keep syncing changes from odh-deployer but mostly from odh-manifests, it might get lost somewhere in the near future when the common monitoring solution is moving into operator entierly

Yes this should be added in the odh-manifests too, great spot thanks Wen! I will update this PR asap. Just having that small issue on scraping KubeRay.

File: monitoring/prometheus/prometheus-configs.yaml (hunk 2 above)
Comment (ChristianZaccaria) [reply monitoring/prometheus/prometheus-configs.yaml]: I can confirm that it is the right namespace

File: monitoring/prometheus/prometheus-configs.yaml (hunk 1 above)
Comment (ChristianZaccaria) [reply monitoring/prometheus/prometheus-configs.yaml:845]: On it now, thanks Wen!

**Comment by zdtsw** [comment]
> > unless i missed something, should not this be added into https://github.com/red-hat-data-services/odh-manifests/blob/master/monitoring/prometheus/prometheus-configs.yaml as well? it wont be in use for v2 for now (since we have not done cloud-service part) we are not keep syncing changes from odh-deployer but mostly from odh-manifests, it might get lost somewhere in the near future when the common monitoring solution is moving into operator entierly
> 
> Yes this should be added in the odh-manifests too, great spot thanks Wen! I will update this PR asap. Just having that small issue on scraping KubeRay.

you dont need to get this into `odh-manifests`, since that repo will be dead before CF with rhods on managed cluster.
as long as the content of this PR is correct, operator will take care of the config.

File: monitoring/prometheus/prometheus-configs.yaml [hunk 3]
```diff
@@ -1144,7 +1046,7 @@ data:
         - role: endpoints
           namespaces:
             names:
-              - redhat-ods-applications
+              - opendatahub
```
Comment (zdtsw) [inline monitoring/prometheus/prometheus-configs.yaml]: i do not think in RHODS we have `opendatahub` namespace

Comment (ChristianZaccaria) [reply monitoring/prometheus/prometheus-configs.yaml]: I double-checked and the CFO and KubeRay will both live in the `redhat-ods-applications` namespace in RHODS. You were right thanks for pointing that out! Changes have been made on latest commit.

**Comment by zdtsw** [comment]
suggest to close this PR, since we are not gonna use odh-deployer for any future development.
ref to new config: https://github.com/red-hat-data-services/rhods-operator/blob/rhods-2.4/config/monitoring/prometheus/apps/prometheus-configs.yaml#L271

**Pull request closed without merging**